server/handlers/clients.js

```diff
@@ -0,0 +1,256 @@
+
+  const clientConfigs = loadClientConfigs(bpanelConfig);
+
+  if (!configsMap) configsMap = createConfigsMap(clientConfigs);
+
+  if (!configsMap.has(id))
+    return res.status(404).json({
+      error: {
+        message: `Sorry, there was no client with the id ${id}`,
+        code: 404
+      }
+    });
+
+  const config = configsMap.get(id);
+
+  assert(config instanceof Config, 'client needs bcfg config');
+
+  const { nodeClient, walletClient, multisigWalletClient } = clientFactory(
+    config
+  );
+
+  const reqClients = {
+    node: nodeClient,
+    wallet: walletClient,
+    multisig: multisigWalletClient
+  };
+
+  const client = reqClients[params.client];
+
+  if (!client)
+    return res.status(404).json({
+      error: {
+        message: `Requested client ${params.client} for ${id} does not exist`,
+        code: 404
+      }
+    });
+
+  /*
+     * this part of the handler is the proxy to the nodes that the clients
+     * are communicating with
+     */
+
+  // use query params for GET request, otherwise use body
+  const payload = method === 'GET' ? query : body;
+  try {
+    logger.debug(
+      `client: ${client.constructor.name}, method: ${method},`,
+      `path: ${path}`
+    );
+    logger.debug('query:', query);
+    logger.debug('body:', body);
+
+    // proxy the token
+    token = client.token;
+    if ('token' in payload) {
+      client.token = payload.token;
+      logger.debug('Using custom client token');
+    }
+    const response = await client.request(method, path, payload);
+    logger.debug('server response:', response ? response : 'null');
+    if (response) return res.status(200).json(response);
+    // return 404 when response is null due to
+    // resource not being found on server
+    return res.status(404).json({ error: { message: 'resource not found' } });
+  } catch (e) {
+    logger.error(`Error querying ${client.constructor.name}:`, e);
+    return res
+      .status(502)
+      .send({ error: { message: e.message, code: e.code, type: e.type } });
+  } finally {
+    // always reassign the original token
+    client.token = token;
+  }
+}
+
+async function getConfigHandler(req, res) {
+  const { logger } = req;
+  let configurations;
+  try {
+    configurations = await getConfig(req.params.id);
+  } catch (e) {
+    logger.error(e);
+    if (e.code === 'ENOENT')
+      return res.status(404).json({
+        error: {
+          message: `Config for "${req.params.id}" not found`,
+          code: 404
+        }
+      });
+    else
+      return res.status(500).json({
+        error: {
+          message: `There was a problem with your request.`,
+          code: 500
+        }
+      });
+  }
+
+  const info = {
+    configs: configurations.data
+  };
+
+  if (req.query.health) {
+    try {
+      logger.info(`Checking status of client "${req.params.id}"...`);
+      const [err, clientErrors] = await testConfigOptions(configurations);
+      if (!err) info.healthy = true;
```

Comment:
Its not clear based on this interface the case in which there is an error but `clientErrors.failed.length` is 0
I make the assumption that there is an invariant such that if there is an error, there will be at least one element in `clientErrors.failed`, but I don't see an obvious guarantee of that here

Reply:
This leads me to believe the logic should account for the case I mentioned above with:

```js
if (!err)
  ...
else
  ...
```

such that the edge case that isn't handled is handled

Reply:
Yeah, I think that's fair. Will update 👍

Reply:
ok, updated